lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,311 @@
+                    freq: int(max_capacities[cpu] * freq / max(freqs))
+                    for freq in freqs
+                }
+                for cpu, freqs in plat_info['freqs'].items()
+        }
+
+    @classmethod
+    def _collect_capacities_flatten(cls, plat_info):
+        capacities = [
+            capa
+            for freq_capas in cls._collect_capacities(plat_info).values()
+            for capa in freq_capas.values()
+        ]
+
+        # Remove the duplicates from the list
+        return sorted(set(capacities))
+
+    @classmethod
+    def _get_bands(cls, capacities):
+        bands = list(zip(capacities, capacities[1:]))
+
+        # Only keep a number of bands
+        nr_bands = cls.NR_PHASES
+        if len(bands) > nr_bands:
+            # Pick the bands covering the widest range of util, since they
+            # are easier to test
+            bands = sorted(
+                bands,
+                key=lambda band: band[1] - band[0],
+                reverse=True
+            )
+            bands = bands[:nr_bands]
+            bands = sorted(bands, key=itemgetter(0))
+
+        return bands
+
+    @classmethod
+    def _get_phases(cls, plat_info):
+        """
+        Returns a list of phases. Each phase being described by a tuple:
+
+          (uclamp_val, util, (band))
+        """
+        capacities = cls._collect_capacities_flatten(plat_info)
+        bands = cls._get_bands(capacities)
+
+        def band_mid(band):
+            return int((band[1] + band[0]) / 2)
+
+        return [
+                (
+                    band_mid(band),
+                    band_mid(band) / 2,
+                    band
+                )
+                for band in bands
+        ]
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        periods = [
+            Periodic(
+                duty_cycle_pct=(util / PELT_SCALE) * 100,  # util to pct
+                duration_s=5,
+                period_ms=cls.TASK_PERIOD_MS,
+                uclamp_min=uclamp_val,
+                uclamp_max=uclamp_val
+            )
+            for uclamp_val, util, expected in cls._get_phases(plat_info)
+        ]
+
+        return {'task': functools.reduce(lambda a, b: a + b, periods)}
+
+    def _get_task_activation_mean_capacities(self, task, start, end):
+        df = self.trace.analysis.tasks.df_task_activation(task)
+        df = df[['active', 'cpu']]
+
+        df_freq = self.trace.analysis.frequency.df_cpus_frequency()
+        df_freq = df_freq[['cpu', 'frequency']]
+        df_freq = df_freq.pivot(index=None,
+                                columns='cpu',
+                                values='frequency').reset_index()
+        df_freq.set_index('Time', inplace=True)
+        df = df.merge(df_freq, how='outer', left_index=True, right_index=True)
+
+        end = end if not np.isnan(end) else df.last_valid_index()
+
+        # Skip the first half to skip any task migration
+        start += (end - start) / 2
```

Comment:
are we sure it's enough ? Since we got the activation dataframe, we can easily check if there is any migration in the window, and if there is any, set the start to the first activation after it.

```
df = self.trace.analysis.tasks.df_task_activation(task)

# First row after each migration
df = df.copy(deep=False)
df['migration'] = df['cpu'] != df['cpu'].shift()
start_after_migration = df[df['migration']].first_valid_index()

if start_after_migration is not None:
	start = start_after_migration
```
Note the shallow copy of the dataframe since we don't modify existing columns but want to add a new one.

If for some reason we expect multiple migrations, we can take the `last_valid_index()`.

Reply:
After many discussions we finally went for the "let's not slice the phase at all" solution.